File: solutions/fsharp/difference-of-squares/2/DifferenceOfSquares.fs
module DifferenceOfSquares

let private square number = pown number 2
let private sumBy mapper number = [1..number] |> Seq.sumBy mapper
let private fork j f g x = j (f x) (g x)

let squareOfSum (number: int): int = number |> (sumBy id >> square)
let sumOfSquares (number: int): int = number |> sumBy square
let differenceOfSquares (number: int): int = number |> fork (-) squareOfSum sumOfSquares

File: solutions/csharp/difference-of-squares/3/DifferenceOfSquares.cs
public static class DifferenceOfSquares
{
    static int Square(int x) => x * x;

    static int SumBy(int count, Func<int, int>? mapper = null, int from = 1) =>
        Enumerable.Range(from, int.Max(0, count)).Sum(mapper ?? (static x => x));
    
    public static int CalculateSquareOfSum(int max) =>
        Square(SumBy(max));

    public static int CalculateSumOfSquares(int max) =>
        SumBy(max, Square);

    public static int CalculateDifferenceOfSquares(int max) =>
        CalculateSquareOfSum(max) - CalculateSumOfSquares(max);
}